src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -840,18 +865,42 @@ void StorageReplicatedMergeTree::dropReplica(zkutil::ZooKeeperPtr zookeeper, con
     ///       "The local set of parts of table X doesn't look like the set of parts in ZooKeeper"
     ///
     {
-        Strings children = zookeeper->getChildren(remote_replica_path);
-
-        if (std::find(children.begin(), children.end(), "metadata") != children.end())
-            zookeeper->remove(fs::path(remote_replica_path) / "metadata");
+        /// Remove metadata first
+        [[maybe_unused]] auto code = zookeeper->tryRemove(fs::path(remote_replica_path) / "metadata");
+        assert(code == Coordination::Error::ZOK || code == Coordination::Error::ZNONODE);
 
-        for (const auto & child : children)
+        /// Then try to remove paths that are known to be flat (all children are leafs)
+        Strings flat_nodes = {"flags", "parts", "queue"};
+        for (const auto & node : flat_nodes)
         {
-            if (child != "metadata")
-                zookeeper->removeRecursive(fs::path(remote_replica_path) / child);
+            bool removed_quickly = zookeeper->tryRemoveChildrenRecursive(fs::path(zookeeper_path) / node, /* probably flat */ true);
+            if (!removed_quickly)
+                LOG_WARNING(logger, "Cannot quickly remove node {} and its children (replica: {}). Will remove recursively.",
```

Comment:
The message is a bit wrong. The children are already removed recursively.

Reply:
But based on the comment below when we try to delete the nodes themselves it says
```
        /// Then try to remove nodes that are known to have no children (and should always exist)
```
We still assume that the children nodes are deleted.
The boolean `removed_quickly` tells us only the information if we didn't use the hint properly (it can be moved to the `tryRemoveChildrenRecursive`) or someone is deleting at the same time but the deletion of child nodes was already done (or at least tried).

Reply:
I've updated the message.
I think it's not a problem if `zookeeper->tryMulti(ops, res)` will fail to remove nodes without children because `tryRemoveChildrenRecursive` left some garbage. Looks like it's extremely rare case.